jsapp/js/universalTable/universalTable.component.tsx

```diff
@@ -0,0 +1,325 @@
+// Libraries
+import React from 'react';
+import cx from 'classnames';
+import {
+  flexRender,
+  getCoreRowModel,
+  useReactTable,
+  type CellContext,
+  type Column,
+  type PaginationState,
+  type TableOptions,
+} from '@tanstack/react-table';
+
+// Partial components
+import Button from 'js/components/common/button';
+import KoboSelect from 'js/components/common/koboSelect';
+
+// Utilities
+import {generateUuid} from 'js/utils';
+
+// Styles
+import styles from './universalTable.module.scss';
+
+export interface UniversalTableColumn {
+  /**
+   * Pairs to data object properties. It is using dot notation, so it's possible
+   * to match data from a nested object :ok:.
+   */
+  key: string;
+  /**
+   * Most of the times this would be just a string, but we are open to
+   * anything really.
+   */
+  label: React.ReactNode;
+  isPinned?: boolean;
+  /**
+   * This is override for the default width of a column. Use it if you need more
+   * space for your data, or if you display something very short.
+   */
+  size?: number;
+  /**
+   * This is an optional formatter function that will be used when rendering
+   * the cell value. Without it a literal text value will be rendered.
+   */
+  cellFormatter?: (value: string) => React.ReactNode;
+}
+
+interface UniversalTableProps<DataItem> {
+  /** A list of column definitions */
+  columns: UniversalTableColumn[];
+  data: DataItem[];
+  // PAGINATION
+  // To see footer with pagination you need to pass all these below:
+  /** Starts with `0` */
+  pageIndex?: number;
+  /** Total number of pages of data. */
+  pageCount?: number;
+  /**
+   * One of `pageSizes`. It is de facto the `limit` from the `offset` + `limit`
+   * pair used for paginatin the endpoint.
+   */
+  pageSize?: number;
+  pageSizes?: number[];
```

Comment:
What would you think about pageSizeOptions as a name for this, just to be a bit more clear?

Reply:
Good idea :ok: